### ISSUE TITLE
 how to replace the wrong node definitions in the frozen graph when importing frozen graph with batchnorm

### ISSUE THREAD

**Issue by bitwanghe** [tensorflow/tensorflow#31524]
there is a confusion when i try to do as the answers in the question #3628, and could anyone give me a proper solution for it?
he says:
''
An ugly way to get it working:
manually replace the wrong node definitions in the frozen graph
RefSwitch --> Switch + add '/read' to the input names
AssignSub --> Sub + remove use_locking attributes
''

but, I printed the nodes in my net,:
''
inputs
labels
keep_prob
MobilenetV2/input
MobilenetV2/Conv/weights/Initializer/truncated_normal/shape
MobilenetV2/Conv/weights/Initializer/truncated_normal/mean
MobilenetV2/Conv/weights/Initializer/truncated_normal/stddev
MobilenetV2/Conv/weights/Initializer/truncated_normal/TruncatedNormal
MobilenetV2/Conv/weights/Initializer/truncated_normal/mul
MobilenetV2/Conv/weights/Initializer/truncated_normal
MobilenetV2/Conv/weights
MobilenetV2/Conv/weights/Assign
MobilenetV2/Conv/weights/read
MobilenetV2/Conv/kernel/Regularizer/l2_regularizer/scale
MobilenetV2/Conv/kernel/Regularizer/l2_regularizer/L2Loss
MobilenetV2/Conv/kernel/Regularizer/l2_regularizer
MobilenetV2/Conv/dilation_rate
MobilenetV2/Conv/Conv2D
MobilenetV2/Conv/BatchNorm/gamma/Initializer/ones
MobilenetV2/Conv/BatchNorm/gamma
MobilenetV2/Conv/BatchNorm/gamma/Assign
MobilenetV2/Conv/BatchNorm/gamma/read
MobilenetV2/Conv/BatchNorm/beta/Initializer/zeros
MobilenetV2/Conv/BatchNorm/beta
MobilenetV2/Conv/BatchNorm/beta/Assign
MobilenetV2/Conv/BatchNorm/beta/read
MobilenetV2/Conv/BatchNorm/moving_mean/Initializer/zeros
MobilenetV2/Conv/BatchNorm/moving_mean
MobilenetV2/Conv/BatchNorm/moving_mean/Assign
MobilenetV2/Conv/BatchNorm/moving_mean/read
MobilenetV2/Conv/BatchNorm/moving_variance/Initializer/ones
MobilenetV2/Conv/BatchNorm/moving_variance
MobilenetV2/Conv/BatchNorm/moving_variance/Assign
MobilenetV2/Conv/BatchNorm/moving_variance/read
MobilenetV2/Conv/BatchNorm/Const
MobilenetV2/Conv/BatchNorm/Const_1
MobilenetV2/Conv/BatchNorm/FusedBatchNorm
MobilenetV2/Conv/BatchNorm/Const_2
MobilenetV2/Conv/BatchNorm/AssignMovingAvg/sub/x
MobilenetV2/Conv/BatchNorm/AssignMovingAvg/sub
MobilenetV2/Conv/BatchNorm/AssignMovingAvg/sub_1
MobilenetV2/Conv/BatchNorm/AssignMovingAvg/mul
MobilenetV2/Conv/BatchNorm/AssignMovingAvg
MobilenetV2/Conv/BatchNorm/AssignMovingAvg_1/sub/x
MobilenetV2/Conv/BatchNorm /AssignMovingAvg_1/sub
MobilenetV2/Conv/BatchNorm/AssignMovingAvg_1/sub_1
MobilenetV2/Conv/BatchNorm/AssignMovingAvg_1/mul
MobilenetV2/Conv/BatchNorm/AssignMovingAvg_1
MobilenetV2/Conv/Relu6
MobilenetV2/expanded_conv/input
''

but! the 'nodes.op' of the 'nodes' above donot exist "RefSwitch", please do you any suggestions?

**Comment by bitwanghe** [comment]
btw, the op.name of
“MobilenetV2/Conv/BatchNorm/AssignMovingAvg”  and "MobilenetV2/Conv/BatchNorm/AssignMovingAvg_1"
are all "assignsub"

**Comment by bitwanghe** [comment]
btw, the op.name of
“MobilenetV2/Conv/BatchNorm/AssignMovingAvg” and "MobilenetV2/Conv/BatchNorm/AssignMovingAvg_1"
are all "assignsub;

**Comment by oanush** [comment]
@bitwanghe

**Comment by bitwanghe** [comment]
> @bitwanghe,
> Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version.

' Win 7 '
' tensorflow-1.13.1 ' 
I restore the parameters from mobilenetV2 , and rewrite the last layer. Then retrain the whole graph.

Thank you!

**Comment by bitwanghe** [comment]
here is my code:
> @bitwanghe,
> Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version.


if __name__ == '__main__':

    print("loading model...")
    tf.reset_default_graph()  # 构建计算图
    os.environ["CUDA_VISIBLE_DEVICES"] = "0"
    finetune_model_dir = './mobilenet_v2_1.0_224/mobilenet_v2_1.0_224.ckpt'
    train_records_dir = './train_records/train_records.txt'
    val_records_dir = './val_records/val_records.txt'
    train_labels_dir = './train_label.txt'
    val_labels_dir = './val_label.txt'

    batch_size = 16
    num_classes = 223
    train_epoches = 1000
    train_num = get_example_nums(train_labels_dir)
    train_iterations = train_num // batch_size
    val_num = get_example_nums(val_labels_dir)
    val_iterations = val_num // batch_size
    lr = 0.01
    val_log_step = 1


    inputs = tf.placeholder(tf.float32, shape=[None, 224, 224, 3], name='inputs')
    labels = tf.placeholder(tf.int32, shape=[None], name='labels')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    #     is_training = tf.placeholder(tf.bool, name='is_training')
    is_training = True

    with slim.arg_scope(mobilenet_v2.training_scope()):
        net0, endpoints = mobilenet_v2.mobilenet(inputs, num_classes=None, is_training=is_training)

     with tf.variable_scope('Logits'):
        logits = slim.conv2d(net0, num_classes, [1, 1], activation_fn=None, normalizer_fn=None, biases_initializer=tf.zeros_initializer(), scope='Conv2d_1c_1x1')
        logits = tf.squeeze(logits, axis=[1, 2])
        

    checkpoint_exclude_scopes = 'Logits'
    exclusions = None
    if checkpoint_exclude_scopes:
        exclusions = [scope.strip() for scope in checkpoint_exclude_scopes.split(',')]
    variables_to_restore = []
    totalvar = slim.get_model_variables()

    for var in slim.get_model_variables():
        excluded = False
        for exclusion in exclusions:
            if var.op.name.startswith(exclusion):
                excluded = True
        if not excluded:
            variables_to_restore.append(var)

    losses = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits)
    loss = tf.reduce_mean(losses)
    logits = tf.nn.softmax(logits, name='probability')
    classes = tf.argmax(logits, axis=1, name='classes')
    accuracy = tf.reduce_mean(tf.cast(tf.equal(tf.cast(classes, dtype=tf.int32), labels), dtype=tf.float32))

    step = tf.get_variable("step", [], initializer=tf.constant_initializer(0.0), trainable=False)
    optimizer = tf.train.AdamOptimizer(learning_rate=lr)
    train_step = slim.learning.create_train_op(loss, optimizer, global_step=step)
    update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
    if update_ops:
        # print("BN parameters: ", update_ops)
        updates = tf.group(*update_ops)
        train_step = control_flow_ops.with_dependencies([updates], train_step)

    saver_restore = tf.train.Saver(var_list=variables_to_restore)
    saver = tf.train.Saver(tf.global_variables(), max_to_keep=2)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        saver_restore.restore(sess, finetune_model_dir)

        train_images_tf, train_labels_tf = read_records(train_records_dir)
        train_images_tf_batch, train_labels_tf_batch = get_batch_images(train_images_tf, train_labels_tf, batch_size, num_classes, shuffle=False)
        val_images_tf, val_labels_tf = read_records(val_records_dir)
        val_images_tf_batch, val_labels_tf_batch = get_batch_images(val_images_tf, val_labels_tf, batch_size, num_classes, shuffle=False)

        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(sess=sess, coord=coord)
        print("start training...")
        for i in range(train_epoches):
            for j in range(train_iterations):
                train_images_batch, train_labels_batch = sess.run([train_images_tf_batch, train_labels_tf_batch])
                train_dict = {inputs: train_images_batch, labels: train_labels_batch, keep_prob: 0.5}
                classes_, labels_, loss_, acc_, _ = sess.run([classes, labels, loss, accuracy, train_step], feed_dict=train_dict)

            if i % val_log_step == 0:
                tmp1 = "%s: train results: Step [%d]  Loss : %f, accuracy :  %g" % (datetime.now(), i, loss_, acc_)
                print(tmp1)

                mean_loss, mean_acc = net_evaluation(sess, loss, accuracy, val_images_tf_batch, val_labels_tf_batch, val_iterations)
                tmp2 = "%s: val results: Step [%d]  Loss : %f,  accuracy :  %g" % ( datetime.now(), i, mean_loss, mean_acc)
                print(tmp2)


            if i % 1 == 0 or (i + 1) == train_epoches:
                savePath = './models/saved_model.ckpt'
                if os.path.exists(savePath) == False:
                    os.mkdir(savePath)
                saver.save(sess, savePath, global_step=i)

        coord.request_stop()
        coord.join(threads)

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
